Firestore/Swift/Source/Result/DocumentReference+Result.swift

```diff
@@ -0,0 +1,266 @@
+  ///   immediately.
+  ///   - result: The result of request. On success it is empty, otherwise it contains an `Error`.
+  /// - Throws: `Firestore.Encoder` encoding errors.
+  func setData<T: Encodable>(from value: T, mergeFields: [Any],
+                             encoder: Firestore.Encoder = Firestore.Encoder(),
+                             completion: @escaping (_ result: Result<Void, Error>)
+                               -> Void) throws {
+    try setData(
+      from: value,
+      mergeFields: mergeFields,
+      encoder: encoder,
+      completion: mapResultCompletion(completion)
+    )
+  }
+
+  /// Updates the document referred to by this `DocumentReference` with the specified data.
+  /// If no document exists, the update fails with an error.
+  /// If a document already exists, it is overwritten.
+  ///
+  /// - Parameters:
+  ///   - fields: The `Dictionary` containing the fields (expressed as `String` or `FieldPath`) and
+  ///   values with which to update the document.
+  ///   - completion: The closure to execute on successfully updating to the server or on receipt of
+  ///   an error.
+  ///   It will not be called while the client is offline, though local changes will be visible
+  ///   immediately.
+  ///   - result: The result of request. On success it is empty, otherwise it contains an `Error`.
+  func updateData(_ fields: [AnyHashable: Any],
+                  completion: @escaping (_ result: Result<Void, Error>) -> Void) {
+    updateData(fields, completion: mapResultCompletion(completion))
+  }
+
+  /// Delete the document referred to by this `DocumentReference` with the specified data.
+  ///
+  /// - Parameters:
+  ///   - completion: The closure to execute on successfully deleting to the server or on receipt of
+  ///   an error.
+  ///   It will not be called while the client is offline, though local changes will be visible
+  ///   immediately.
+  ///   - result: The result of request. On success it is empty, otherwise it contains an `Error`.
+  func delete(completion: @escaping (_ result: Result<Void, Error>) -> Void) {
+    delete(completion: mapResultCompletion(completion))
+  }
+
+  /// Reads the documents matching this query.
+  ///
+  /// - Parameters:
+  ///   - source: The `FirestoreSource` where the results should be fetched:
+  ///     - `default`:  Fetch from the server and, if it fails, the cache.
+  ///     - `server`:  Fetch from the server only.
+  ///     - `cache`:  Fetch from the cache only.
+  ///   - completion: The closure to execute on receipt of a result.
+  ///   - result: The result of request. On success it contains the `DocumentSnapshot`, otherwise an
+  ///   `Error`.
+  func getDocument(source: FirestoreSource = .default,
+                   completion: @escaping (_ result: Result<DocumentSnapshot, Error>) -> Void) {
+    getDocument(source: source, completion: mapResultCompletion(completion))
+  }
+
+  /// Attaches a listener for this `DocumentReference` events.
```

Comment:
The reason I changed `DocumentSnapshot` to `DocumentReference` was that int the corresponding `Query` method says: "Attaches a listener for this `Query` events"... it seemed an inconsistency to me since:
`Query` <-> `DocumentReference`
`QuerySnapshot` <-> `DocumentSnapshot`

Reply:
I'm waiting for a response at the first point and a definitive description of all `result`s. All other issues are fixed.

Reply:
Thanks for calling this out. I think your description is very close to what I would put, I think instead of:
```
result: A DocumentSnapshot if Firestore has successfully received a an update, otherwise an Error. If Firestore delivers an Error, the listener is automatically removed.
```

something to better indicate it will likely be called multiple times:

```
result: A `DocumentSnapshot` when Firestore successfully receives updates, otherwise an error. If an Error occurs, the listener is automatically removed.
```

What do you folks think?

Reply:
I started thinking that saying _"result of request"_ or _"Firestore successfully receives updates"_ could be misleading since Firestore could return a cached result.
What about this (based on [this](https://developer.apple.com/documentation/swift/result/success) description):
```
result: Stores a `DocumentSnapshot`, otherwise an error. If an error occurs, the listener is automatically removed.
```

Reply:
Re https://github.com/firebase/firebase-ios-sdk/pull/6940/files#r524768134: this really should be `DocumentSnapshot`. The `Query` method should refer to `QuerySnapshot` events rather than `Query` events.

Re https://github.com/firebase/firebase-ios-sdk/pull/6940/files#r526849654: "stores" is probably not the right verb to use here since Firestore nominally stores data and we're reading here. Also, "on success" is probably too ambiguous about what's going on. How about:

```
result: A `DocumentSnapshot` when Firestore has an update (from the cache or server), otherwise an error. If an error occurs, the listener is automatically removed.
```

?

Reply:
@wilhuff 
> Re https://github.com/firebase/firebase-ios-sdk/pull/6940/files#r524768134: this really should be DocumentSnapshot. The Query method should refer to QuerySnapshot events rather than Query events.

Ok, I get it.

> Re https://github.com/firebase/firebase-ios-sdk/pull/6940/files#r526849654: "stores" is probably not the right verb to use here since Firestore nominally stores data and we're reading here.

Stores was referred to the `Result` enum, but I understand the ambiguity. 

>  Also, "on success" is probably too ambiguous about what's going on. How about:
> 
>     result: A `DocumentSnapshot` when Firestore has an update (from the cache or server), otherwise an error. If an error occurs, the listener is automatically removed.
> ?

I found a little oddity in the sentence. An update come only from the server, an initial state can come from the cache. At the limit a local change for the latency compensation can come from the cache, but I'm not sure if it's really coming from cache. If is this the case I would propose to change `has an update` with `gets a value`.
In any case what should be put in the `getDocument` `Result` descriptions?

Reply:
Correcting this:
```suggestion
  /// Attaches a listener for this `DocumentSnapshot` events.
```